src/Altinn.App.Core/EFormidling/Implementation/EformidlingStatusCheckEventHandler.cs

```diff
@@ -29,15 +30,45 @@ public class EformidlingStatusCheckEventHandler : IEventHandler
         private readonly IEFormidlingClient _eFormidlingClient;
         private readonly ILogger<EformidlingStatusCheckEventHandler> _logger;
         private readonly IHttpClientFactory _httpClientFactory;
-        private readonly IMaskinportenService _maskinportenService;
-        private readonly MaskinportenSettings _maskinportenSettings;
-        private readonly IX509CertificateProvider _x509CertificateProvider;
+        private readonly IMaskinportenTokenProvider _maskinportenTokenProvider;
         private readonly PlatformSettings _platformSettings;
         private readonly GeneralSettings _generalSettings;
 
+        [Obsolete("Do not use this for new code. Will be removed as part of V9.")]
+        private readonly IMaskinportenService? _maskinportenService;
+        [Obsolete("Do not use this for new code. Will be removed as part of V9.")]
+        private readonly MaskinportenSettings? _maskinportenSettings;
+        [Obsolete("Do not use this for new code. Will be removed as part of V9.")]
+        private readonly IX509CertificateProvider? _x509CertificateProvider;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EformidlingStatusCheckEventHandler"/> class.
+        /// </summary>
+        public EformidlingStatusCheckEventHandler(
+            IEFormidlingClient eFormidlingClient,
+            IHttpClientFactory httpClientFactory,
+            ILogger<EformidlingStatusCheckEventHandler> logger,
+            IMaskinportenTokenProvider? maskinportenTokenProvider,
+            IOptions<PlatformSettings> platformSettings,
+            IOptions<GeneralSettings> generalSettings
+            )
+        {
+            // Even though maskinportenTokenProvider is nullable, it is required for the new constructor.
+            // This is to allow for a smooth transition from the old to the new constructor.
+            ArgumentNullException.ThrowIfNull(maskinportenTokenProvider);
+
+            _eFormidlingClient = eFormidlingClient;
+            _logger = logger;
+            _httpClientFactory = httpClientFactory;
+            _maskinportenTokenProvider = maskinportenTokenProvider;
+            _platformSettings = platformSettings.Value;
+            _generalSettings = generalSettings.Value;
+        }
+
         /// <summary>
         /// Initializes a new instance of the <see cref="EformidlingStatusCheckEventHandler"/> class.
         /// </summary>
+        [Obsolete("Use the constructor using IMaskinportenTokenProvider instead. This constructor will be removed when V9 is released.")]
```

Comment:
You are right that this compatibility relies on either the Jwk or X509 provider being registered, but not both. If the both are it will use the X509 as this has the most parameters. This was what I mentioned in the description of this PR.

But I agree, I think I overcomplicated this by trying to keep the same EformidlingStatusCheckEventHandler class and could only have created a new version of it using the IMaskinportenTokenProvider.